### PR TITLE
The great shared_ptr purge

### PR DESCRIPTION
**Pull request by jokva**
Hi,

For a long time we've talked about getting rid of the pointless shared_ptr use in opm-parser, and now most of them are gone. This has a few implications worth mentioning:

* I've unearthed some bugs because of this
* Deck is now copyable, and EclipseState and friends will hopefully be soon.
* Copying generally now means copying
* Storage type is no longer mandated by opm-parser; the decision is left to the caller.
* Interfaces are clearer i.e. they don't imply ownership sharing anymore, which was never really the intention.

This PR does not really change any behaviour. However, downstream code WILL break from both interface changes and the removal of all the *Ptr aliases. Downstream PRs will be coming.

### CODE REVIEW

**Comment by atgeirr** [comment]
I think this is good, and you are free to merge this across all modules of which I am maintainer once all tests have run.

**Comment by atgeirr** [comment]
jenkins build this opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 run norne spe polymer with downstream please

**Comment by atgeirr** [comment]
@akva2 could you take a look at the Jenkins command I issued and verify that it indeed should build and run absolutely everything? It's the longest I have written...

**Comment by akva2** [comment]
it will do most things but you cannot do run triggers from other places than opm-simulators (only its scripts parse it).

**Comment by atgeirr** [comment]
> you cannot do run triggers from other places than opm-simulators

Thanks for reminding me! In the notes at

http://wiki.opm-project.org/index.php?title=Jenkins_triggers

I had written that you can do this from opm-data as well is that still true?

In any case, I started a Jenkins job from the opm-simulators PR to get the run tests.

**Comment by akva2** [comment]
sorry for being imprecise. yes, you can trigger them from opm-data in addition to opm-simulators.

**Comment by jokva** [comment]
Mind running it again with material included?

**Comment by akva2** [comment]
you can trigger here yourself fyi.

**Comment by atgeirr** [comment]
> Mind running it again with material included?

I am not sure that would help, it looks like nothing downstream was triggered. Perhaps I did something wrong in the trigger. In any case, I'll not start it here but in opm-simulators.

**Comment by akva2** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by atgeirr** [comment]
BTW, the previous run in opm-simulators dis not fail in opm-material as I would expect, but with this:
```
/var/lib/jenkins/workspace/opm-simulators-PR-builder/deps/opm-parser/opm/parser/eclipse/python/c_inter/cparser.cc:19:24: error: 'const class Opm::Parser' has no member named 'newDeckFromFile'
         return parser->newDeckFromFile( file , *parse_mode );
```

**Comment by akva2** [comment]
please no comments before jenkins starts to see if it is the old bug that bites or not.

**Comment by akva2** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by akva2** [comment]
yup. it is the stupid bug in the pr builder. sometimes (i haven't found the cause), it will pass the wrong comment from github if comments are added between trigger and poll cycle. but it's not consistent, sometimes it's okay..

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by akva2** [comment]
still going to fail with ```CMake Error at cmake_install.cmake:60 (file):
  file INSTALL cannot find
  "/var/lib/jenkins/workspace/opm-parser-PR-builder/deps/opm-output/opm/output/eclipse/CornerpointChopper.hpp".```

**Comment by jokva** [comment]
Hm. That seems like a jenkins detail that will be fixed by cleaning out the workspace, no?

**Comment by akva2** [comment]
jenkins always run on a clean workspace. that doesn't help when you instruct the build system to install an (now) nonexistent file :)

**Comment by jokva** [comment]
You're absolutely right. I hadn't gotten it included in the commit, but the line itself was gone from my file.

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by akva2** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by joakim-hove** [comment]
OK - this is very good; should obviously be merged. Would like to merge: opm/opm-output#101 first. Depending on how things pan out this might come in immediately after the release.

**Comment by atgeirr** [comment]
> Would like to merge: OPM/opm-output#101 first

I agree, probably less risk of rebasing errors doing it in that direction.

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
Unless someone screams this is going in once Jenkins is done.

**Comment by atgeirr** [comment]
> Unless someone screams this is going in once Jenkins is done.

Should be fine, but please verify first that the Jenkins run did what we expect it to, and that all the merge buttons are green.

**Comment by jokva** [comment]
Affirmative.

**Comment by jokva** [comment]
Eh, and how do I do that?

**Comment by akva2** [comment]
by looking at the log (job sits in queue, i'll lend you a pair of eyes once it has run).

**Comment by jokva** [comment]
I'm afraid I have no idea where the jenkins logs reside, but thanks, that's appreciated.

**Comment by akva2** [comment]
under the checks there is a 'details' link once the job has started. that takes you to the jenkins job, where you can look at the test results, the console output etc.

**Comment by jokva** [comment]
Oh, that's neat. I'll keep a look out and also be a little bit patient.

**Comment by akva2** [comment]
failure in opm-core, you might have to rebase the pr.

**Comment by jokva** [comment]
Already had, but must've not pushed it.

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

File: opm/parser/eclipse/Units/Units.hpp [hunk 1]
```diff
@@ -248,7 +248,9 @@ namespace Opm {
         namespace detail {
             // Some units are sometimes unused, and generate a (potentially) large number of warnings
             // Adding them here silences these warnings, and should have no side-effects
-            constexpr double __attribute__((unused)) unused_units = stb + liter + barsa + psia + darcy;
```
Comment (andlaus) [inline opm/parser/eclipse/Units/Units.hpp]: better remove the whole `#if [...] #endif` branch...

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by atgeirr** [comment]
Following Jenkins, I see it got past opm-grid! Perhaps we finally got this...

**Comment by atgeirr** [comment]
> Perhaps we finally got this...

Just to have hopes squashed immediately after. opm-simulators stuff now...

**Comment by jokva** [comment]
Ugh, more MPI stuff. I'll rebuild simulators and purge the rest.

**Comment by atgeirr** [comment]
@jokva Do you have an MPI/Zoltan stack now for testing, so it does not take 1 hour to get feedback via Jenkins for every error?

**Comment by jokva** [comment]
Yes.

**Comment by blattms** [comment]
Well, there is definitely a bug in PR of opm-grid. I commented on the
diff there. That bug will not be triggered by Jenkins.

**Comment by jokva** [comment]
Simulator has been fixed, upscaling doesn't seem to have that funky code paths with MPI.

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by GitPaean** [comment]
The compilation of the OPM chain (no ewoms) looks fine from my side.

**Comment by andlaus** [comment]
> The compilation of the OPM chain (no ewoms) looks fine from my side. 

don't worry about ewoms (and the frankenstein branch of `opm-simulators` for this matter): If it breaks because of this series, I'll fix it ASAP.

**Comment by jokva** [comment]
... except for the benchmark bit of upscaling. Which has been fixed.

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 please

**Comment by jokva** [comment]
jenkins build this with downstreams opm-output=109 opm-core=1089 opm-simulators=863 opm-grid=242 opm-upscaling=206 opm-material=170 ewoms=73 please

**Comment by jokva** [comment]
I guess building with a patched ewoms help.

I just built zoltan+mpi enabled ewoms here (I don't think it matters, but better safe than sorry) and it compiled fine, so I expect this one to go through.

**Comment by jokva** [comment]
There we go, successful build.

**Pull request merged**